src/bentoml/server.py

```diff
@@ -0,0 +1,274 @@
+from __future__ import annotations
+
+import sys
+import typing as t
+import logging
+import subprocess
+from abc import ABC
+from abc import abstractmethod
+from typing import TYPE_CHECKING
+
+from simple_di import inject
+from simple_di import Provide
+
+from ._internal.tag import Tag
+from ._internal.bento import Bento
+from ._internal.configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    from types import TracebackType
+
+    from ._internal.client import Client
+    from ._internal.client import GrpcClient
+    from ._internal.client import HTTPClient
+
+
+logger = logging.getLogger(__name__)
+
+
+class Server(ABC):
+    bento: str | Bento | Tag
+    host: str
+    port: int
+
+    args: list[str]
+
+    process: subprocess.Popen[bytes] | None = None
+    timeout: int = 10
+    _client: Client | None = None
+
+    def __init__(
+        self,
+        bento: str | Bento | Tag,
+        serve_cmd: str,
+        reload: bool,
+        production: bool,
+        env: t.Literal["conda"] | None,
+        host: str,
+        port: int,
+        working_dir: str | None,
+        api_workers: int | None,
+        backlog: int,
+    ):
+        self.bento = bento
+
+        if isinstance(bento, Bento):
+            bento_str = str(bento.tag)
+        elif isinstance(bento, Tag):
+            bento_str = str(bento)
+        else:
+            bento_str = bento
+
+        args: list[str] = [
+            sys.executable,
+            "-m",
+            "bentoml",
+            "serve-http",
+            bento_str,
+            "--host",
+            host,
+            "--port",
+            str(port),
+            "--backlog",
+            str(backlog),
+        ]
+
+        if production:
+            args.append("--production")
+        if reload:
+            args.append("--reload")
+        if env:
+            args.extend(["--env", env])
+
+        if api_workers is not None:
+            args.extend(["--api-workers", str(api_workers)])
+        if working_dir is not None:
+            args.extend(["--working-dir", str(working_dir)])
+
+        self.args = args
+        self.host = "127.0.0.1" if host == "0.0.0.0" else host
```

Comment:
What's the reason for this conversion from "0.0.0.0" to "127.0.0.1"?

Reply:
0.0.0.0 isn't a valid IP address, it's just means bind all IPs, so we convert it to something sensible.